aws_lambda_powertools/event_handler/middlewares/openapi_validation.py

```diff
@@ -0,0 +1,329 @@
+    _regenerate_error_with_loc,
+    get_missing_field_error,
+)
+from aws_lambda_powertools.event_handler.openapi.encoders import jsonable_encoder
+from aws_lambda_powertools.event_handler.openapi.exceptions import RequestValidationError
+from aws_lambda_powertools.event_handler.openapi.params import Param
+from aws_lambda_powertools.event_handler.openapi.types import IncEx
+from aws_lambda_powertools.event_handler.types import EventHandlerInstance
+
+logger = logging.getLogger(__name__)
+
+
+class OpenAPIValidationMiddleware(BaseMiddlewareHandler):
+    """
+    OpenAPIValidationMiddleware is a middleware that validates the request against the OpenAPI schema defined by the
+    Lambda handler. It also validates the response against the OpenAPI schema defined by the Lambda handler. It
+    should not be used directly, but rather through the `enable_validation` parameter of the `ApiGatewayResolver`.
+
+    Examples
+    --------
+
+    ```python
+    from typing import List
+
+    from pydantic import BaseModel
+
+    from aws_lambda_powertools.event_handler.api_gateway import (
+        APIGatewayRestResolver,
+    )
+
+    class Todo(BaseModel):
+      name: str
+
+    app = APIGatewayRestResolver(enable_validation=True)
+
+    @app.get("/todos")
+    def get_todos(): List[Todo]:
+      return [Todo(name="hello world")]
+    ```
+    """
+
+    def handler(self, app: EventHandlerInstance, next_middleware: NextMiddleware) -> Response:
+        logger.debug("OpenAPIValidationMiddleware handler")
+
+        route: Route = app.context["_route"]
+
+        values: Dict[str, Any] = {}
+        errors: List[Any] = []
+
+        try:
+            # Process path values, which can be found on the route_args
+            path_values, path_errors = _request_params_to_args(
+                route.dependant.path_params,
+                app.context["_route_args"],
+            )
+
+            # Process query values
+            query_values, query_errors = _request_params_to_args(
+                route.dependant.query_params,
+                app.current_event.query_string_parameters or {},
+            )
+
+            values.update(path_values)
+            values.update(query_values)
+            errors += path_errors + query_errors
+
+            # Process the request body, if it exists
+            if route.dependant.body_params:
+                (body_values, body_errors) = _request_body_to_args(
+                    required_params=route.dependant.body_params,
+                    received_body=self._get_body(app),
+                )
+                values.update(body_values)
+                errors.extend(body_errors)
+
+            if errors:
+                # Raise the validation errors
+                raise RequestValidationError(_normalize_errors(errors))
+            else:
+                # Re-write the route_args with the validated values, and call the next middleware
+                app.context["_route_args"] = values
```

Comment:
This is what allows us to write a handler like this `def handler(user: User, id: int)...`